cmd/heal.go

```diff
@@ -0,0 +1,359 @@
+	ServerSideEncryption encrypt.ServerSide `json:"sse"`
+}
+
+const timeFormat = "2006-01-02T15:04:05.000Z"
+
+func errLogName() string {
+	return fmt.Sprintf("error_%s.log", time.Now().Format(timeFormat))
+}
+
+func newErrorFile(fname string) (f *os.File, err error) {
+	if fname != "" {
+		f, err = os.OpenFile(fname, os.O_CREATE|os.O_WRONLY|os.O_SYNC, 0666)
+		if err != nil {
+			return nil, err
+		}
+		return
+	}
+	return
+}
+
+// HealSys struct represents the config for healing system
+type HealSys struct {
+	ctx     context.Context
+	Dir     string
+	errorCh chan ErrorLog
+	healCh  chan ErrorLog
+}
+
+// newHealSys initializes healing
+func newHealSys(ctx context.Context, dir string) *HealSys {
+	if dir == "" {
+		return nil
+	}
+	sys := &HealSys{ctx: ctx, Dir: dir, errorCh: make(chan ErrorLog, 10000), healCh: make(chan ErrorLog, 10000)}
+	go sys.heal(ctx)
+	return sys
+}
+
+func (sys *HealSys) init() {
+	if sys == nil {
+		return
+	}
+	if err := os.MkdirAll(sys.Dir, 0777); err != nil {
+		logger.FatalIf(err, fmt.Sprintf("Unable to initialize '%s' dir", sys.Dir))
+	}
+	files, err := readDir(sys.Dir)
+	if err != nil {
+		logger.FatalIf(err, fmt.Sprintf("Invalid error log dir %s", sys.Dir))
+	}
+	for _, fname := range files {
+		sys.process(sys.ctx, fname)
+	}
+	go sys.startErrLogger(sys.ctx)
+}
+
+// process sends error logs to the heal channel for an attempt to heal.
+func (sys *HealSys) process(ctx context.Context, fname string) {
+	fpath := pathJoin(sys.Dir, fname)
+	file, err := os.Open(fpath)
+	if err != nil {
+		logger.LogIf(ctx, err)
+		return
+	}
+	defer file.Close()
+	bio := bufio.NewScanner(file)
+	for bio.Scan() {
+		var errLog ErrorLog
+		if err = msgpack.Unmarshal(bio.Bytes(), &errLog); err != nil {
+			// TODO: decide on retry
+			continue
+		}
+		sys.healCh <- errLog
+	}
+	if err = bio.Err(); err != nil {
+		logger.LogIf(ctx, err)
+	}
+	os.Remove(fpath)
+}
+
+// send error log to error channel
+func (sys *HealSys) send(ctx context.Context, log ErrorLog) {
+	if sys == nil {
+		return
+	}
+	select {
+	case sys.errorCh <- log:
+		return
+	case <-ctx.Done():
+		return
+	}
+}
+
+// startErrLogger logs radio API call errors to a temporary file for further processing
+// by healing routine.
+func (sys *HealSys) startErrLogger(ctx context.Context) {
+	prevLogName := ""
+	currLogName := errLogName()
+	fptr, err := newErrorFile(pathJoin(sys.Dir, currLogName))
+	if err != nil {
+		logger.LogIf(ctx, err)
+	}
+	for {
+		select {
+		case log, ok := <-sys.errorCh:
+			if !ok {
+				continue
+			}
+			for {
+				bytes, err := msgpack.Marshal(&log)
+				if err != nil {
+					logger.LogIf(ctx, err)
+				}
+
+				if _, err = fptr.Write(bytes); err != nil {
+					logger.LogIf(ctx, err)
+					continue
+				}
+				fptr.WriteString("\n")
+				break
+			}
+		case <-time.After(30 * time.Second):
+			fptr.Close()
+			prevLogName = currLogName
+			currLogName = errLogName()
+			fptr, err = newErrorFile(pathJoin(sys.Dir, currLogName))
```

Comment:
This is a weird model we should have a better way to close and re-open a file this is super racy..

Reply:
To avoid significant opens and re-opens lets actually keep the file pointer open even for processing, I am still confused about this part of the code though let me think about it more.